app/services/page_cloner.rb

```diff
@@ -0,0 +1,79 @@
+class PageCloner
+  def self.clone!(page)
+    new(page).clone!
+  end
+
+  def initialize(page)
+    @page = page
+  end
+
+  def clone!
+    cloned_page = @page.dup
+
+    # TODO: I'm thinking to have the user of this API
+    # decide what the new title should be and have it
+    # passed in to clone! It's not this class' responsibility
+    # to do the necessary checks against AK
+    cloned_page.title = "#{@page.title} #{Time.now.to_s}"
+
+    cloner(cloned_page) do |clone|
+      clone.links
+      clone.plugins
+      clone.tags
+      clone.images
+    end
+
+    cloned_page.save
+    cloned_page
+  end
+
+  private
+
+  def cloner(cloned_page)
+    machine = Cloner.new(cloned_page, @page)
+    yield(machine)
```

Comment:
The block is defined in `clone!`

```ruby
cloner do
  links
  plugins
  tags
  images
end
```

I'm only using one here to keep the public API as readable and declarative as possible.